linker-private-library/src/main/scala/org/scalajs/linker/runtime/RyuDouble.scala

```diff
@@ -0,0 +1,1162 @@
+    }
+    val sign = bits < 0
+
+    /* Step 2: Determine the interval of information-preserving outputs.
+     * Compute smaller halfway point, f, and larger halfway point as
+     * u*2^e2, v*2^e2, and w*2^e2. Where e2 = e - 2 to make sure u,v,w are unsigned integers.
+     *
+     * Given f- and f+ are smaller and larger floating point number to f,
+     * (f + f-)/2 = (m*2^e + (m-1)*2^e)/2 = m*2^e - (1/2)*2^e = (4m - 2)*e2.
+     * Therefore, u=4m-2, but if the mantissa is zero, the next smaller numbers are
+     * m'=(1...1), e'=e-1, and resulting in (4m-1)*e2
+     *
+     *    f-                f                 f+
+     * ---|--------|--------|--------|--------|--->
+     *          u*2^e2   v*2^e2   w*2^e2
+     */
+    val even = (m & 1) == 0
+    /* Is it "general case" for calculating the lower boundary u.
+     * This will be false if mantissa is all zeros, and the exponent > 1.
+     */
+    val isGeneralLowerBoundCase = (m != 1L << MantissaBits) || exponent <= 1
+    val u = 4 * m - (if (isGeneralLowerBoundCase) 2 else 1)
+    val v = 4 * m
+    val w = 4 * m + 2
+    var e2 = e - 2
+
+    /* Step 3: Convert the binary representation (u,v,w) * 2^e2 to a
+     * decimal representation (a,b,c) * 10^e10.
+     *
+     * We also combine the main Ryu optimization (Step 4' in the paper)
+     * by removing 'q' digits in one efficient step. This is done by
+     * multiplying with precomputed values, which keeps the result
+     * within the information-preserving interval.
+     *
+     * The result of removing 'q' digits is (aq,bq,cq), which
+     * represents the final interval (aq,bq,cq) * 10^e10.
+     *
+     * The returning flags flags (za, zb, and zc) track if the 'q' digits
+     * removed during the conversion step were all zeros.
+     */
+    val (aq, bq, cq, e10, za, zb) =
+      convertBinaryToDecimalInterval(u, v, w, e2, isGeneralLowerBoundCase, even)
+
+    // Step 4: Find the shortest decimal representation in the interval of legal representations.
+    val vplength = decimalLength(cq)
+    val exp = e10 + vplength - 1
+
+    /* Per ECMA-262 Number::toString specification:
+     * Decimal notation is used if the exponent n is in the [-5, 21],
+     * otherwise, scientific notation must be used.
+     * Here, n in the ECMA spec equals to the `exp + 1`,
+     * and the scientific notation condition is `!(exp ∈ [-6, -20])`.
+     */
+    val scientificNotation = (exp < -6) || (exp > 20)
+    val (output, removed) = computeShortest(aq, bq, cq, za, zb, e10, even, scientificNotation)
+    val olength = vplength - removed
+
+    // Step 5: Print the decimal representation.
+    // We follow Double.toString semantics here.
+    printString(output, olength, exp, sign, scientificNotation)
+  }
+
+  @inline
+  private def convertBinaryToDecimalInterval(u: Long, v: Long, w: Long, e2: Int,
+      isGeneralLowerBoundCase: Boolean, even: Boolean): (Long, Long, Long, Int, Boolean, Boolean) = {
```

Comment:
Ouch, this shouldn't work. I think the `JavalibIRCleaner` is too lenient in the linker-private-library. It shouldn't allow references to our custom tuple classes from the javalib. The linker-private-library should always be able to link with any ABI-compliant version of the javalib.

Use a custom `@inline private final class` instead of the 6-tuple.